### PR TITLE
Add possibility to change legacy LV2 port value by the plugin's DSP part

### PR DESCRIPTION
**Pull request by sadko4u**
The value of this feature: the plugin may introduce non-standard control modes like MIDI control, remote OSC control or something similar. In this case, it is not possible to synchronize changed legacy port parameters against Host, DSP and UI without having proper interface to pass the new value of the legacy parameter.

Modern plugin formats like CLAP and VST3 allow to automate the changes of the plugin's parameters from the audio thread by issuing corresponding messages or filling corresponding data structures. Even modern hosts accept the VST2's `audioMasterAutomate` callback issued by the DSP part of the plugin code - [example in JUCE framework](https://github.com/juce-framework/JUCE/blob/4ada2e1f4db082f5dd803f249f37b3bf9c5f55ef/modules/juce_audio_plugin_client/juce_audio_plugin_client_VST2.cpp#L698-L707). According to my current knowledge, there is no such feature in LV2 yet.

This PR aims to introduce such feature with minimal changes by allowing to pass legacy LV2 port index in `patch:Set` messages issued by the plugin.
In this case the host is responsible for handling this message and changing the value of the legacy LV2 port on the next `run` call. The host is also permitted to reject the requested value in case it is playing automation or handling user input.
To synchronize the changed value with plugin's UI, the host should transfer the new value via the `ui:PortNotification` message.

### CODE REVIEW

**Comment by sadko4u** [comment]
There are no implementations. But there is a request to add remote OSC control to LSP Plugins. With LV2 I found that it will be a problem to properly synchronized DSP and UI if some OSC control is performed with closed UI in the DAW.
I think @x42 and @falkTX are interested, too.

**Comment by x42** [comment]
Ardour has some 8+ year old code to allow a plugin to automate itself. 
http://lists.lv2plug.in/pipermail/devel-lv2plug.in/2016-February/001571.html
In that case the action has to be initiated by a user and the changes are timestamped.

Other than that I still think it's a bad idea to allow a plugin to change its own inputs. 
Control Values are for a user to set and should not change behind a user's back.

AFAIK VST3 does also not allow this.

**Comment by falkTX** [comment]
> AFAIK VST3 does also not allow this.

VST3 has parameter output events which does exactly this

**Comment by sadko4u** [comment]
> Ardour has some 8+ year old code to allow a plugin to automate itself.

Is it possible to look at the headers?

**Comment by sadko4u** [comment]
> Those are for read-only parameters (kIsReadOnly, not kCanAutomate)

I don't recall I've ever seen this restriction mentioned in the VST3 SDK source code.

[nor here](https://github.com/steinbergmedia/vst3_pluginterfaces/blob/151ecde4d6ee1c457dcce848342b162461944fe6/vst/ivstparameterchanges.h)
[nor here](https://github.com/steinbergmedia/vst3_pluginterfaces/blob/151ecde4d6ee1c457dcce848342b162461944fe6/vst/ivstaudioprocessor.h)

As I also see, [for JUCE it doesn't matter](https://github.com/juce-framework/JUCE/blob/4ada2e1f4db082f5dd803f249f37b3bf9c5f55ef/modules/juce_audio_plugin_client/juce_audio_plugin_client_VST3.cpp#L3606-L3619) if the parameter is `kReadOnly`.

**Comment by x42** [comment]
But [here](https://forums.steinberg.net/t/clarification-of-parameter-handling-in-vst-3/201914) 

... when an audio processor wants to change a parameter value it can use IConnectionPoint to send a message to the edit controller, but should not do so on the real-time audio thread...

**Comment by x42** [comment]
IOW the `IParamValueQueue` used as outputParameterChanges` process data ***should*** only contain data for read-only ports

**Comment by sadko4u** [comment]
By the other side, [the answer](https://forums.steinberg.net/t/clarification-of-parameter-handling-in-vst-3/201914/2) is:

> Yes, it is one way to do it. An another way, for example for VuMeter parameter, the plugin could export the parameter as normal parameter and send them back in the process call into the outputParameterChanges queue. The host will them at the right time to the edit Controller (see Again example).

IOW there is a mechainsm of changing the parameter value and it should be done by the controller.
Controller in VST3 is not a GUI and co-exists with the Processor.
In LV2 we don't have any controller. Actually the Host is a controller.

**Comment by drobilla** [comment]
> Other than that I still think it's a bad idea to allow a plugin to change its own inputs

I agree this is a bad idea in a vacuum, as are pretty much all ideas related to kludging around control ports.  This is "dealing with legacy" stuff.  It is, however, an, er, _unoffensive_ addition, spec-wise.  Plugins can already trivially send such a message, and standardized messages are better than not.  It's just an extension of an existing mechanism to be able to refer to (&#%@!) port indices.

The main thing that's missing is a Feature to describe it and/or detect support (generally lacking with message stuff to be fair), but I need to look at the schema changes more closely first...

**Comment by sadko4u** [comment]
Hi @drobilla! Look at the latest FabFilter Pro-Q4 release with multiple instances control in a single plugin view.
Without possibility of changing legacy LV2 parameters from the DSP side such multi-instancing solution looks not possible for LV2.

**Comment by sadko4u** [comment]
Hello! Any news with this?

**Comment by drobilla** [comment]
Please don't abuse tickets to nag me.  No, there isn't any news.  If there was, then you'd be able to see it.

As far as I know, there is no host implementation of this.  It's not a particularly high priority thing for me personally to implement.  If you want this feature as a plugin developer, then it's your responsibility to work with a host to get it implemented.

**Comment by sadko4u** [comment]
I think, the problem can be viewed from another side: no one developer of the host won't implement this feature until it is not available in standard specification.

**Comment by sadko4u** [comment]
@falkTX did you have any real application cases of the extension you've introduced in Carla?
What do you think about my variation of implementing such feature introduced with this PR?

@drobilla I'm not trying to nag you rather I'm trying to force the LV2 stadard to be on par with modern CLAP and VST3 formats and not to loose functionality. At least lacking some functionality in LV2 makes impossible to develop plugins that support multiple formats.

**Comment by falkTX** [comment]
> @falkTX did you have any real application cases of the extension you've introduced in Carla?

I have use for it from within the MOD platform, that is where the need from it came from.
When changing some params we wanted to have that disable some other options.

> What do you think about my variation of implementing such feature introduced with this PR?

I dont like your approach because it requires the use of atom ports, which is overkill for simple lv2 plugins.
If a plugin has atom ports, it might as well be using patch parameters already too.

**Comment by drobilla** [comment]
> @drobilla I'm not trying to nag you

Great.  In the future then, please pay attention to whether your comments add any useful information to the conversation.  If they don't, then they are "nagging".  For example:

> Hi, @drobilla. Any news?

This is nagging.  Then there was a more appropriate comment that actually contained some information.  Then,

> Hello! Any news with this?

Which is also nagging.  You may not like my word choice, but it's objectively so: there's no content other than to try and push me into doing work you want me to do (for free, on your timeline).  Stop it.  I'm aware that there are things to do - I'm very aware that there are a ton of utterly thankless things to do.  If you want things implemented in some host somewhere and fully incorporated into the standard in a releasable way quickly, _then do the work_.

Now I see there's fundamental disagreement even between active plugin developers?  Clearly this isn't appropriate for me to just blindly ram in the standard.  What you're asking me to do isn't a simple maintenance task, or something just to do with the LV2 project itself, but to spearhead an entire effort across the ecosystem to get this standardized.  Well, there's a pretty big list of efforts I could spearhead across the LV2 ecosystem, and kludges around legacy control ports aren't very high on the list of ones that I actually want to do.

I'm happy to do my "job" as maintainer and incorporate low-hanging fruit, but this isn't that.  If you want things incorporated quickly, you need to make them low-hanging fruit.  Aside from trivial additions, that means general agreement (or at least not strong opposition) and at the very least two implementations that get along with each other.  What I see here isn't that, rather, _all_ of the feedback that I'm aware of (x42 and falkTX in this thread) is negative _and_ there's no working implementations.  On top of that, the whole problem is rooted in using legacy control ports (at a time when there's finally been some motion in actively getting away from them no less).

I'm not about to blindly ram something like that in the spec to strong-arm implementations to (maybe) later implement it (if the idea turns out to even work at all).  That's not how any of this works.

**Comment by sadko4u** [comment]
@drobilla the last [comment](https://github.com/lv2/lv2/pull/72#issuecomment-2386785205) from your side dated Oct, 1st, was very promising and looked like you will get involved into the problem in the nearest future.
So here are almost four months passed and there is no solution. I would be glad if you even rejected this PR but proposed your own solution.

> If you want things implemented in some host somewhere and fully incorporated into the standard in a releasable way quickly, then do the work.

This is a wrong approach. That's why we already have three or four different implementations of the same thing. The lack in standard forces people to do their own implementation comfortable for them.
For example, I don't aggree with @falkTX solution but understand the reason he did so.

> Now I see there's fundamental disagreement even between active plugin developers? 

This is normal. See above. The product with long support latency is often considered a dead project.

> What you're asking me to do isn't a simple maintenance task, or something just to do with the LV2 project itself, but to spearhead an entire effort across the ecosystem to get this standardized.

As a conuterpart I can take as example the [CLAP](https://github.com/free-audio/clap/issues) repository. There are many proposals, many experimental things that become stable after some releases. I don't see CLAP developers suffering from standardizing new things.

> Aside from trivial additions, that means general agreement (or at least not strong opposition) and at the very least two implementations that get along with each other. 

I thought it is a role of maintainer to get the final working solution. We see a problem, we need to solve it. The standard shouldn't depend only on the developers of hosts. If LV2 standard won't be on par with other modern standards, then it will be declined by plugin developers.

> On top of that, the whole problem is rooted in using legacy control ports (at a time when there's finally been some motion in actively getting away from them no less).

Legacy control ports existed, exist and will exist. People will prefer legacy control ports to the patch properties because they are much more natural and have constraints which is not true about patch properties. How many DAWs support patch properties in their UIs? And how many DAWs support legacy properties? Huh.

And, again. I do the work. But I'm a plugin developer, not a host developer. By the other side, if you have not much time to support LV2 standard, it is a good reason to widen the number of developers who will deal with it's opportune support. At least @x42 and @falkTX are good candidates, arent't they?

**Comment by drobilla** [comment]
> I would be glad if you even rejected this PR

Okay.

**Comment by sadko4u** [comment]
But you didn't propose your solution.

**Comment by sadko4u** [comment]
> Conveniently though, one of the main ideas behind LV2 is that things don't need to be in the main specification/distribution anyway, so you're free to publish it on your own.

And we will have three, four or ten implementations of the same thing that plugin developers and host developers should be optionally supporting. While the standard strictly tells the only way to do some thing right. LV2 already looks like over-engineered standard with sharp entry threshold, now it additionally will look much more unpleasant for potential developers.  OKay.

> Sure didn't. I won't reward this kind of toxic and entitled behaviour. Words apparently didn't work, so hopefully a demonstration will. This feature is no longer on my personal TODO list.

Oh man, that looks crazy. Tell anyone that the maintainer of LV2 standard rejects to improve it because he got offended - won't believe.

If you think that you're the only busy one, you're wrong. I also have many things to do but that doesn't make me to not to respond people for months and keep PRs and small improvements not resolved for years. I'm unhappy with such toxic dictatorship maintenance of LV2. Sad but it looks like I have to accept that the standard will die because of its creator.

**Comment by drobilla** [comment]
> And we will have three, four or ten implementations of the same thing that plugin developers and host developers should be optionally supporting.

This is already the case.  You jamming _yet another one_ into the standard doesn't make the others go away, it makes the situation worse.

> LV2 already looks like over-engineered standard with sharp entry threshold

Sure does.  Something you're trying to _make even worse_ by hastily jamming even more unimplemented crap into it.

> I'm unhappy with such toxic dictatorship maintenance of LV2

... you're literally upset because it's _not_ a toxic dictatorship that conveniently does what you want, which in this case is to steamroll other existing implementations with your unimplemented idea, despite the fact that literally every other developer that has weighed in doesn't like it.  Yes, what a horribly toxic dictatorship :roll_eyes:

That said, I certainly _can_ be dictatorial, and since you're apparently incapable of being less unhinged (even a whole day later), it looks like I might have to.  Consider this your last warning: any further toxic debate bro behaviour like this and you'll be blocked from the project indefinitely.

**Comment by thoth-amon** [comment]
> ... you're literally upset because it's not a toxic dictatorship that conveniently does what you want

He's upset because (from his perspective) you stopped dead a conversation and then dressed him down for "nagging" when he was trying to get an update. That was unnecessary, "Sorry, I've been busy" would have been less confrontational. Both of you then went over the top at each other in the ensuing back and forth.  You are both doing important and great projects to linux audio that are related, so from a user perspective, can you reset your relations, sort out better communications, have a voice/video chat.  Going at each other over text is foolish.

**Comment by drobilla** [comment]
Some passerby apparently thought it a good idea to resurrect a toxic thread and encourage the (notoriously combative, unreasonable, and very much part of a long-standing pattern) bad behaviour that ran things off the rails in the first place.  In order to prevent further escalation, I've nuked the entirety of the thread.  Any similar issues in the future will be dealt with in the same way.

Nothing of value was lost here, but in the interest of basic record-keeping, I'll provide a summary of why this PR is closed without all of the confusion and drama:

The only thing that really matters here is this: implementation is king.  By design, anyone is free to implement nearly anything they like within LV2.  If they want things to be in The Standard, awesome, that's generally best, however they need to be implemented at least once before that happens (even a half-assed implementation will do).  I could maybe be budged on that, but at the absolute bare minimum there needs to be a decent discussion about it with rationale and buy-in from _someone_.  These are perfectly reasonable guidelines.  I doubt that any sensible person seriously thinks these aren't reasonable guidelines, but if they do, they're free to do it in their own project instead.  It's not even necessary to make a break with LV2 in order to do so, again, by design.

In summary:

1. Nobody implemented this.
2. Alternative approaches to solve the same problem _are_ implemented.
3. 100% of the other developers involved in the conversation disagreed with the proposal, and their concerns weren't addressed (literally not addressed at all, not even in discussion).

_Obviously_ that's not suitable for inclusion.  So, since it was explicitly demanded that I either include it immediately or close the issue, the issue is closed.
